### PR TITLE
Migrate-crypto-rebase

### PR DESCRIPTION
**Pull request by wolverineks**

### CODE REVIEW

File: src/Settings/BiometricsLink/BiometricsLinkScreen.tsx [hunk 1]
```diff
@@ -18,7 +18,8 @@ export const BiometricsLinkScreen = () => {
   const dispatch = useDispatch()
   const linkBiometricsSignIn = async () => {
     if (await canBiometricEncryptionBeEnabled()) {
-      dispatch(setSystemAuth(true))
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      ;(dispatch(setSystemAuth(true)) as any)
```
Comment (stackchain) [inline src/Settings/BiometricsLink/BiometricsLinkScreen.tsx:22]:
```suggestion
      (dispatch(setSystemAuth(true)) as any)
```

File: src/components/ConfirmTx/ConfirmTx.test.tsx [hunk 2]
```diff
@@ -3,10 +3,11 @@ import {render} from '@testing-library/react-native'
 import React from 'react'
 import {Provider} from 'react-redux'
 
-import configureStore from '../../../legacy/helpers/configureStore.js'
+import configureStore from '../../legacy/configureStore.js'
 import {HaskellShelleyTxSignRequest} from '../../yoroi-wallets/index.js'
 
-global.window = {}
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+;(global as any).window = {}
```
Comment (stackchain) [inline src/components/ConfirmTx/ConfirmTx.test.tsx:10]:
```suggestion
(global as any).window = {}
```

File: src/components/PinInput/PinInput.tsx [hunk 3]
```diff
@@ -3,8 +3,8 @@ import React from 'react'
 import {Image, TouchableHighlight, View} from 'react-native'
 import {StyleSheet} from 'react-native'
 
-import backspaceIcon from '../../../legacy/assets/img/backspace.png'
-import utfSymbols from '../../../legacy/utils/utfSymbols'
+import backspaceIcon from '../../assets//img/backspace.png'
```
Comment (stackchain) [inline src/components/PinInput/PinInput.tsx]:
```suggestion
import backspaceIcon from '../../assets/img/backspace.png'
```

File: src/components/PinInputKeyboard/PinInputKeyboard.tsx [hunk 4]
```diff
@@ -1,8 +1,8 @@
 import React, {useState} from 'react'
 import {Image, StyleSheet, TouchableHighlight, View} from 'react-native'
 
-import backspaceIcon from '../../../legacy/assets/img/backspace.png'
-import utfSymbols from '../../../legacy/utils/utfSymbols'
+import backspaceIcon from '../../assets//img/backspace.png'
```
Comment (stackchain) [inline src/components/PinInputKeyboard/PinInputKeyboard.tsx]:
```suggestion
import backspaceIcon from '../../assets/img/backspace.png'
```

File: src/legacy/ledgerUtils.ts [hunk 5]
```diff
@@ -58,22 +56,20 @@ import TransportHID from '@v-almonacid/react-native-hid'
 import {PermissionsAndroid, Platform} from 'react-native'
 import {BleError} from 'react-native-ble-plx'
 
+import {ledgerMessages} from '../../legacy/i18n/global-messages'
+import LocalizableError from '../../legacy/i18n/LocalizableError'
+import {Logger} from '../../legacy/utils/logging'
 // $FlowExpectedError
```
Comment (stackchain) [inline src/legacy/ledgerUtils.ts]:
```suggestion
```

File: src/yoroi-wallets/cardano/metadataUtils.ts [hunk 6]
```diff
@@ -27,14 +24,16 @@ export async function createAuxiliaryData(auxiliary: Array<JSONMetadata>): Auxil
   return await AuxiliaryData.new(metadata)
 }
 
-export function parseMetadata(hex: string): any {
+export async function parseMetadata(hex: string) {
   const metadatum = TransactionMetadatum.from_bytes(Buffer.from(hex, 'hex'))
```
Comment (stackchain) [inline src/yoroi-wallets/cardano/metadataUtils.ts]:
```suggestion
  const metadatum = await TransactionMetadatum.from_bytes(Buffer.from(hex, 'hex'))
```

File: src/yoroi-wallets/cardano/metadataUtils.ts [hunk 7]
```diff
@@ -27,14 +24,16 @@ export async function createAuxiliaryData(auxiliary: Array<JSONMetadata>): Auxil
...
-  const metadataString = decode_metadatum_to_json_str(metadatum, MetadataJsonSchema.BasicConversions)
+  const metadataString = decode_metadatum_to_json_str(await metadatum, MetadataJsonSchema.BasicConversions)
```
Comment (stackchain) [inline src/yoroi-wallets/cardano/metadataUtils.ts]:
```suggestion
  const metadataString = decode_metadatum_to_json_str(metadatum, MetadataJsonSchema.BasicConversions)
```

File: src/yoroi-wallets/cardano/metadataUtils.ts [hunk 8]
```diff
@@ -27,14 +24,16 @@ export async function createAuxiliaryData(auxiliary: Array<JSONMetadata>): Auxil
...
+
   return metadataString
 }
 
-export function parseMetadataDetailed(hex: string): any {
+export async function parseMetadataDetailed(hex: string) {
   const metadatum = TransactionMetadatum.from_bytes(Buffer.from(hex, 'hex'))
```
Comment (stackchain) [inline src/yoroi-wallets/cardano/metadataUtils.ts]:
```suggestion
  const metadatum = await TransactionMetadatum.from_bytes(Buffer.from(hex, 'hex'))
```

File: src/yoroi-wallets/cardano/metadataUtils.ts [hunk 9]
```diff
@@ -27,14 +24,16 @@ export async function createAuxiliaryData(auxiliary: Array<JSONMetadata>): Auxil
...
-  const metadataString = decode_metadatum_to_json_str(metadatum, MetadataJsonSchema.DetailedSchema)
+  const metadataString = decode_metadatum_to_json_str(await metadatum, MetadataJsonSchema.DetailedSchema)
```
Comment (stackchain) [inline src/yoroi-wallets/cardano/metadataUtils.ts]:
```suggestion
  const metadataString = decode_metadatum_to_json_str(metadatum, MetadataJsonSchema.DetailedSchema)
```

File: src/yoroi-wallets/cardano/shelley/delegationUtils.ts [hunk 10]
```diff
@@ -23,39 +21,45 @@ import {BigNumber} from 'bignumber.js'
...
+import {CardanoError, InsufficientFunds, RewardAddressEmptyError} from '../../../legacy/errors'
+import {ObjectValues} from '../../../legacy/flow'
+import type {DefaultAsset} from '../../../legacy/HistoryTransaction'
+import type {CardanoHaskellShelleyNetwork} from '../../../legacy/networks'
+import type {BackendConfig} from '../../../legacy/types'
+import type {AccountStateRequest, AccountStateResponse} from '../../../legacy/types'
+import type {AddressedUtxo, Addressing, V4UnsignedTxAddressedUtxoResponse} from '../../../legacy/types'
+import {normalizeToAddress} from '../../../legacy/utils'
+import {StakingStatus} from '../../../types'
 // $FlowExpectedError
```
Comment (stackchain) [inline src/yoroi-wallets/cardano/shelley/delegationUtils.ts]:
```suggestion
```

File: src/yoroi-wallets/utils/parsing.ts [hunk 11]
```diff
@@ -14,26 +13,26 @@ export class InvalidAssetAmount extends ExtendableError {
...
+    LT_MIN_UTXO: 'LT_MIN_UTXO',
+    // amount is less than min utxo value allowed
     NEGATIVE: 'NEGATIVE',
   }
 
-  constructor(errorCode: $Values<typeof InvalidAssetAmount.ERROR_CODES>) {
+  constructor(errorCode: typeof InvalidAssetAmount.ERROR_CODES[keyof typeof InvalidAssetAmount.ERROR_CODES]) {
     super('InvalidAssetAmount')
-    this.errorCode = errorCode
+    ;(this as any).errorCode = errorCode
```
Comment (stackchain) [inline src/yoroi-wallets/utils/parsing.ts:23]:
```suggestion
    (this as any).errorCode = errorCode
```

File: tsconfig.json [hunk 12]
```diff
@@ -24,6 +24,6 @@
     "resolveJsonModule": true,
     "skipLibCheck": true
   },
-  "include": ["./.d.ts", "./src"],
+  "include": ["./.d.ts", "./src", "legacy/actions/tokenInfo.js"],
```
Comment (stackchain) [inline tsconfig.json]: it might be added automatically

Comment (stackchain) [inline tsconfig.json]:
```suggestion
  "include": ["./.d.ts", "./src"],
```

File: src/Settings/BiometricsLink/BiometricsLinkScreen.tsx (hunk 1 above)
Comment (wolverineks) [reply src/Settings/BiometricsLink/BiometricsLinkScreen.tsx:22]: the semicolon is added automatically, it cant be removed

File: src/components/ConfirmTx/ConfirmTx.test.tsx (hunk 2 above)
Comment (wolverineks) [reply src/components/ConfirmTx/ConfirmTx.test.tsx:10]: the semicolon is added automatically, it cant be removed

File: src/yoroi-wallets/utils/parsing.ts (hunk 11 above)
Comment (wolverineks) [reply src/yoroi-wallets/utils/parsing.ts:23]: the semicolon is added automatically, it cant be removed

**Review by stackchain — CHANGES_REQUESTED** [review]

**Pull request merged**